### PR TITLE
Fonts 2024 Queries

### PR DESCRIPTION
**Pull request by IvanUkhov**
In this PR, the queries from 2022 are copied to 2024 and are updated as follows:

## Changes

* The data source is exclusively `httparchive.all`.
* All queries are annotated with the question they answer.
* All queries follow the same structure and style.
* Queries are split by the section in the chapter where they are used.
* Repeated parts are extracted into common functions.

## Structure

The queries are split by the section where they are used:

* `design/` is about foundries and families,
* `development/` is about tools and technologies, and
* `performance/` is about hosting and serving.

Each file name starts with one of the following prefixes indicating the primary subject of the corresponding analysis:

* `fonts_` is about font files,
* `pages_` is about HTML pages,
* `scripts_` is about JavaScript scripts, and
* `styles_` is about CSS style sheets.

The prefix is followed by the property studied given in singular, potentially extended one or several suffixes narrowing down the scope, as in `fonts_size_by_table.sql` and `pages_link_relation.sql`.

## Contents

Each query starts with a preamble indicating the section, question, and normalization type:

```sql
-- Section: Performance
-- Question: What is the distribution of the file size broken down by table?
-- Normalization: Websites
```

Many queries rely on temporary functions for convenience and clarity. The functions appear in several queries are extracted into a common file: `common.sql`. Whenever any of the functions defined in `common.sql` is used by a query, the query has the following line at the top:

```sql
-- INCLUDE https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2024/fonts/common.sql
```

It signalizes that, prior to executing the query, `common.sql` has to be inlined.

### CODE REVIEW

File: sql/2024/fonts/development/fonts_variable_by_service.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
+-- Section: Development
+-- Question: Who is serving variable fonts?
+
+CREATE TEMPORARY FUNCTION SERVICE(url STRING) AS (
+  CASE
+    WHEN REGEXP_CONTAINS(url, r'((use|fonts)\.typekit\.(net|com))|webfonts\.creativecloud\.com') THEN 'Adobe'
```
Comment (IvanUkhov) [inline sql/2024/fonts/development/fonts_variable_by_service.sql:6]: @bramstein, is the last one attributed correctly? It was under `edgefonts.net` in 2022.

Comment (IvanUkhov) [reply sql/2024/fonts/development/fonts_variable_by_service.sql:6]: And in general, does this list need any update since 2022?

Comment (bramstein) [reply sql/2024/fonts/development/fonts_variable_by_service.sql:6]: Edge Fonts is a separate (free) web font service by Adobe, but it has been deprecated. As far as I know it still serves fonts however.

I'm not aware of any new web font service used at a scale significant enough that it would show up here.

Comment (IvanUkhov) [reply sql/2024/fonts/development/fonts_variable_by_service.sql:6]: I see. Where should I put `webfonts.creativecloud.com`? Is it part of Edge Fonts?

Comment (IvanUkhov) [reply sql/2024/fonts/development/fonts_variable_by_service.sql:6]: Or we can group Edge Fonts with Adobe if appropriate.

File: sql/2024/fonts/performance/fonts_service.sql
Comment (bramstein) [inline sql/2024/fonts/performance/fonts_service.sql:1]: In the 2022 chapter we showed a graph of service usage over time:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2022/fonts/font_usage_by_service.sql

I think that would be interesting to show as it seems there has been some change in 2023.

File: sql/2024/fonts/performance/fonts_service.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    date = '2024-07-01' AND
+    type = 'font'
+  GROUP BY
+    client,
+    service
+)
+
+SELECT
+  client,
+  service,
```
Comment (bramstein) [inline sql/2024/fonts/performance/fonts_service.sql:35]: The 2022 query also included data on whether a page used a combination of self-hosted or hosted fonts. Can we include that again? It can paint a very different picture...

File: sql/2024/fonts/common.sql [hunk 3]
```diff
@@ -0,0 +1,94 @@
+-- Normalize a name. Used in FAMILY.
+CREATE TEMPORARY FUNCTION FAMILY_INNER(name STRING) AS (
+  IF(LENGTH(TRIM(name)) < 3, NULL, NULLIF(TRIM(name), ''))
+);
+
+-- Extract the family name from a payload.
+CREATE TEMPORARY FUNCTION FAMILY(payload STRING) AS (
+  FAMILY_INNER(JSON_EXTRACT_SCALAR(payload, '$._font_details.names[1]'))
```
Comment (bramstein) [inline sql/2024/fonts/common.sql:8]: We should probably pull this from the CSS as in 2022, as it is very common to remove the name table entries for web fonts (i.e. why a lot of the entries in the data are empty).

File: sql/2024/fonts/common.sql [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+);
+
+-- Extract the file format from a URL and a Content-Type header.
+CREATE TEMPORARY FUNCTION FILE_FORMAT(url STRING, header STRING) AS (
+  LOWER(COALESCE(
```
Comment (bramstein) [inline sql/2024/fonts/common.sql:13]: This is different from previous years, what's the rationale for the change? Are they functionally equivalent? If not, we should make a note in the chapter explaining why it has been changed.

Comment (bramstein) [reply sql/2024/fonts/common.sql:13]: The percentages are quite different from 2022, so we should definitely look into it.

File: sql/2024/fonts/performance/pages_link_relationship.sql
Comment (bramstein) [inline sql/2024/fonts/performance/pages_link_relationship.sql:1]: This should check if the resource hint is for a font, as resource hints can be used for multiple resources. If I'm reading this correctly, this just summarises all resource hint usage instead of all resource hint usage for fonts.

File: sql/2024/fonts/performance/styles_font_display.sql
Comment (bramstein) [inline sql/2024/fonts/performance/styles_font_display.sql:1]: Would love to have the 2023 data here; seems there has been a big change...

File: sql/2024/fonts/performance/scripts_font_face.sql
Comment (bramstein) [inline sql/2024/fonts/performance/scripts_font_face.sql:1]: Would be interesting to get the 2023 data as well. I don't think we included this in the chapter in 2022, but if there's been a significant shift it would be interesting to include.

File: sql/2024/fonts/common.sql (hunk 3 above)
Comment (IvanUkhov) [reply sql/2024/fonts/common.sql:8]: I did not change this aspect. There are queries where the name is taken from the CSS, and there are queries where the name is taken from the naming table. This function is used in the latter queries.

File: sql/2024/fonts/common.sql (hunk 4 above)
Comment (IvanUkhov) [reply sql/2024/fonts/common.sql:13]: You mean this, right?

https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2022/fonts/font_format_usage.sql#L5

There is no `mimeType` in `requests` anymore, but there are request headers. So I take `Content-Type`:

https://github.com/IvanUkhov/almanac.httparchive.org/blob/iu-fonts/sql/2024/fonts/performance/fonts_format_file.sql#L18-L19

Comment (IvanUkhov) [reply sql/2024/fonts/common.sql:13]: And there is no more `ext` either. So I take the show URL.

https://github.com/IvanUkhov/almanac.httparchive.org/blob/iu-fonts/sql/2024/fonts/performance/fonts_format_file.sql#L8

Comment (IvanUkhov) [reply sql/2024/fonts/common.sql:13]: Added two pages showing the same two queries but with the original regular expression. Please take a look:

Skärmavbild 2024-08-12 kl  09 43 00

Comment (JannisBush) [reply sql/2024/fonts/common.sql:13]: @IvanUkhov 
`mimeType` and `ext` still exists, it moved to `summary` in requests: https://har.fyi/reference/structs/request-summary/
So yo can do `JSON_VALUE(summary, '$.mimeType') as mimeType` to access the mimeType and any other value of `summary`.

Comment (IvanUkhov) [reply sql/2024/fonts/common.sql]: Thank you, Jannis! Totally missed that. I have now updated both the two queries and the two pages in the spreadsheet. I will also update the queries regarding the file size; I also took that from the headers. Instead of the summary field.

File: sql/2024/fonts/performance/pages_link_relationship.sql
Comment (IvanUkhov) [reply sql/2024/fonts/performance/pages_link_relationship.sql:1]: Yes, I removed that on purpose. Here is the original query:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2022/fonts/resource_hints_usage.sql

The way I interpret is that it extracts all hints from all pages and then zooms in on those pages where there are fonts, which is probably not what you have in mind.

Comment (IvanUkhov) [reply sql/2024/fonts/performance/pages_link_relationship.sql:1]: Maybe we want to then update the logic by strictly focusing on the HTML content and picking only those hints where `as="font"` or something?

File: sql/2024/fonts/performance/fonts_service.sql
Comment (IvanUkhov) [reply sql/2024/fonts/performance/fonts_service.sql:1]: I have added data for July 2022 and July 2023. The table was not backfilled for dates prior to 2022.

File: sql/2024/fonts/performance/fonts_service.sql (hunk 2 above)
Comment (IvanUkhov) [reply sql/2024/fonts/performance/fonts_service.sql]: I have updated the query to show the actual combinations. Please take a look.

File: sql/2024/fonts/common.sql (hunk 3 above)
Comment (IvanUkhov) [reply sql/2024/fonts/common.sql:8]: More specifically, it is used in the following queries, which are based solely on the font files:

- design/fonts_family_by_foundry.sql,
- design/fonts_family_by_script.sql,
- development/fonts_color_family.sql,
- development/fonts_variable_family.sql, and
- performance/fonts_family_by_service.sql.

File: sql/2024/fonts/performance/pages_link_relationship.sql
Comment (IvanUkhov) [reply sql/2024/fonts/performance/pages_link_relationship.sql:1]: Pushed a change to check `as` as well as `href` in the case of `dns-prefetch` and `preconnect`.

File: sql/2024/fonts/performance/scripts_font_face.sql
Comment (IvanUkhov) [reply sql/2024/fonts/performance/scripts_font_face.sql:1]: Added July 2022 and July 2023.

**Comment by tunetheweb** [comment]
What's the difference between this one and #3667 ?

**Comment by IvanUkhov** [comment]
As I explained there, I did not notice there was a PR open and created a new one. And even if I did, I would probably have to do that anyway to have more control. In that other PR, the queries from 2022 are copied over to the folder for 2024 with no change. So there is no progress there.

File: sql/2024/fonts/performance/styles_font_display.sql
Comment (IvanUkhov) [reply sql/2024/fonts/performance/styles_font_display.sql:1]: Turns out `parsed_css` was not backfilled either. So it gave no data for 2022 and 2023.

Skärmavbild 2024-08-15 kl  10 39 43

File: sql/2024/fonts/design/fonts_foundry.sql [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    rank <= 100
+),
+pages AS (
+  SELECT
+    client,
+    COUNT(DISTINCT page) AS total
+  FROM
+    `httparchive.all.requests`
+  WHERE
+    date = '2024-07-01'
```
Comment (mgifford) [inline sql/2024/fonts/design/fonts_foundry.sql:31]: Are we using the '2024-07-01' or '2024-06-01'?

Comment (IvanUkhov) [reply sql/2024/fonts/design/fonts_foundry.sql:31]: The former for fonts. We did some crawler changes after the latter.

File: sql/2024/fonts/performance/fonts_size.sql [hunk 6]
```diff
@@ -0,0 +1,33 @@
+-- Section: Performance
+-- Question: What is the distribution of the file size?
+
+WITH
+fonts AS (
+  SELECT
+    client,
+    url,
+    AVG(PARSE_NUMERIC(JSON_EXTRACT_SCALAR(summary, '$.respBodySize'))) AS size
```
Comment (bramstein) [inline sql/2024/fonts/performance/fonts_size.sql:9]: We should probably take format into account here as just using WOFF vs. WOFF2 would throw off comparisons.

Comment (bramstein) [reply sql/2024/fonts/performance/fonts_size.sql:9]: (and in the other file size related queries as well)

File: sql/2024/fonts/performance/fonts_format_file.sql [hunk 7]
```diff
@@ -0,0 +1,28 @@
+-- Section: Performance
+-- Question: Which file formats are used?
+
+-- INCLUDE ../common.sql
+
+SELECT
```
Comment (bramstein) [inline sql/2024/fonts/performance/fonts_format_file.sql:6]: The numbers are *very* different from 2022, has anything else changed?

File: sql/2024/fonts/development/fonts_compiler.sql [hunk 8]
```diff
@@ -0,0 +1,48 @@
+-- Section: Design
+-- Question: Which compilers are used?
+
+CREATE TEMPORARY FUNCTION COMPILER(version STRING) AS (
+  CASE
+    WHEN REGEXP_CONTAINS(version, r'(?i)(Core \d|PS \d|hotconv|makeotf)') THEN 'Adobe Font Development Kit for OpenType'
+    WHEN REGEXP_CONTAINS(version, r'(?i)FontCreator') THEN 'FontCreator'
+    WHEN REGEXP_CONTAINS(version, r'(?i)Fontself') THEN 'Fontself Maker'
+    WHEN REGEXP_CONTAINS(version, r'(?i)ttfautohint') THEN 'FreeType'
```
Comment (bramstein) [inline sql/2024/fonts/development/fonts_compiler.sql:9]: I don't think we should attribute ttfautohint to FreeType, as neither are font compilers and the actual "compiling" has been done by one of the other tools.

File: sql/2024/fonts/development/fonts_compiler.sql [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    WHEN REGEXP_CONTAINS(version, r'(?i)(FEAKit|Glyphs)') THEN 'Glyphs.app'
+    WHEN REGEXP_CONTAINS(version, r'(?i)gftools') THEN 'Google Fonts Tools'
```
Comment (bramstein) [inline sql/2024/fonts/development/fonts_compiler.sql:11]: Maybe this should be fontmake?

File: sql/2024/fonts/development/styles_font_variant.sql [hunk 10]
```diff
@@ -0,0 +1,71 @@
+-- Section: Development
+-- Question: Which features are used via font-variant in CSS?
+
+CREATE TEMPORARY FUNCTION PROPERTIES(json STRING)
```
Comment (bramstein) [inline sql/2024/fonts/development/styles_font_variant.sql:4]: The previous version used to compare `font-variant` settings vs. `font-feature-settings`. The `font-feature-settings` data seems to be missing.

Comment (bramstein) [reply sql/2024/fonts/development/styles_font_variant.sql:4]: Nevermind, I see that is now in `style_feature_control.sql`.

File: sql/2024/fonts/development/fonts_compiler.sql (hunk 8 above)
Comment (IvanUkhov) [reply sql/2024/fonts/development/fonts_compiler.sql:9]: Hm, that is true. Removed.

File: sql/2024/fonts/development/fonts_compiler.sql (hunk 9 above)
Comment (IvanUkhov) [reply sql/2024/fonts/development/fonts_compiler.sql:11]: Renamed. Renamed also “Adobe Font Development Kit” to simply “makeotf.”

File: sql/2024/fonts/performance/fonts_format_file.sql (hunk 7 above)
Comment (IvanUkhov) [reply sql/2024/fonts/performance/fonts_format_file.sql:7]: Changed the normalization to “Fonts on pages.”

File: sql/2024/fonts/performance/styles_font_display.sql
Comment (IvanUkhov) [reply sql/2024/fonts/performance/styles_font_display.sql:1]: You mean this, right?

Skärmavbild 2024-09-05 kl  14 34 50

Comment (IvanUkhov) [reply sql/2024/fonts/performance/styles_font_display.sql:1]: `swap` is now around 45%, and `block` 23%. It was noted in 2022, that the growth was 11%.

**Comment by JannisBush** [comment]
Are all the `force-pushes` intentionally?
image
Clicking on my notifications I see this which is a bit annoying.

File: sql/2024/fonts/performance/fonts_size.sql [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+SELECT
+  client,
+  percentile,
+  COUNT(0) AS count,
+  ROUND(APPROX_QUANTILES(size, 1000)[OFFSET(percentile * 10)]) AS size
+FROM
+  fonts,
+  UNNEST([10, 25, 50, 75, 90, 99]) AS percentile
```
Comment (bramstein) [inline sql/2024/fonts/performance/fonts_size.sql:29]: @IvanUkhov We used `10, 25, 50, 75, 90, 100` in the 2022 query. I think this is better though, but we should make a note of it in the chapter.

Comment (IvanUkhov) [reply sql/2024/fonts/performance/fonts_size.sql:29]: Yeah, the 100th percentile is kind of meaningless. It just picks up outliers, which one is not going to show in the report anyway. I am even regretting I did not go lower to, say, 95 in all queries calculating percentiles.

Comment (tunetheweb) [reply sql/2024/fonts/performance/fonts_size.sql:29]: We normally show 10, 25, 50, 75, 90. The 100 is for the top stat which can some times make for a funny throw away line (“Shockingly the largest amount of fonts loaded by a single page is 123,456”) but we don’t normally include them in the actual charts (in part because they often skew the access).

Comment (bramstein) [reply sql/2024/fonts/performance/fonts_size.sql:29]: Yea, we didn't use the 100th percentile in the 2022 article, but it was in the spreadsheet. I was just pointing out the difference in the query as we were internally discussing the difference between the 2022 data and 2024 data which I think can be explained by this change.

Comment (tunetheweb) [reply sql/2024/fonts/performance/fonts_size.sql:29]: Well, given my understanding for what this is used for in the past, I'm not sure 99 (or 95) is better than 100. But  up to you all!

File: sql/2024/fonts/README.md [hunk 12]
```diff
@@ -1,20 +1,54 @@
...
+
+```sql
+-- INCLUDE ../common.sql
+```
 
-- [📄 Planning doc][~google-doc]
-- [📊 Results sheet][~google-sheets]
-- [📝 Markdown file][~chapter-markdown]
+It signalizes that, prior to executing the query, `common.sql` has to be
+inlined.
```
Comment (tunetheweb) [inline sql/2024/fonts/README.md:50]: The queries are linked from the Web Almanac and this won’t be understood.

I think we should remove the common.sql files and just inline them in the files.

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: sql/2024/fonts/common.sql [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+AS r"""
+if (codepoints && codepoints.length) {
+  return detectWritingScript(codepoints.map((character) => parseInt(character, 10)), 0.05);
+} else {
+  return [];
+}
+""";
+
+-- Infer scripts from a payload.
+CREATE TEMPORARY FUNCTION SCRIPTS(payload STRING) AS (
```
Comment (tunetheweb) [inline sql/2024/fonts/common.sql:61]: I worry about people confusing scripts with JavaScript (I think I did earlier in the chat for one!). Should we call this something else? WRITING_SCRIPT? FONT_SCRIPT? CODEPOINTS?

File: sql/2024/fonts/design/fonts_family_by_script.sql [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    FAMILY(payload) AS family,
+    COUNT(0) OVER (PARTITION BY client) AS total
+  FROM
+    `httparchive.all.requests`
+  WHERE
+    date = '2024-07-01' AND
+    type = 'font' AND
+    is_root_page AND
+    IS_PARSED(payload) AND
+    RAND() > 0.5
```
Comment (tunetheweb) [inline sql/2024/fonts/design/fonts_family_by_script.sql:21]: Why? Can you add a comment to explain?

File: sql/2024/fonts/design/fonts_family_by_script.sql [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+SELECT
+  client,
+  script,
+  family,
+  2 * COUNT(0) AS count,
+  2 * total AS total,
```
Comment (tunetheweb) [inline sql/2024/fonts/design/fonts_family_by_script.sql:29]: Why? Can you add a comment to explain (same for line above).

File: sql/2024/fonts/design/fonts_designer.sql [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+designers AS (
+  SELECT
+    client,
+    NULLIF(TRIM(JSON_EXTRACT_SCALAR(payload, '$._font_details.names[9]')), '') AS designer,
+    COUNT(DISTINCT page) AS count,
+    ROW_NUMBER() OVER (PARTITION BY client ORDER BY COUNT(DISTINCT page) DESC) AS rank
+  FROM
+    `httparchive.all.requests`
+  WHERE
+    date = '2024-07-01' AND
```
Comment (tunetheweb) [inline sql/2024/fonts/design/fonts_designer.sql:17]: I thought we were using 2024-06-01 for web Almanac. Is there a reason we can’t use this?

File: sql/2024/fonts/design/fonts_metric.sql [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+  percentile,
+  COUNT(0) AS count,
+  APPROX_QUANTILES(
+    IF(name = 'granularity', value, SAFE_DIVIDE(value, granularity)),
+    1000
+  )[OFFSET(percentile * 10)] AS value
+FROM
+  fonts,
+  UNNEST(metrics) AS metric,
+  UNNEST([10, 25, 50, 75, 90, 99]) AS percentile
```
Comment (tunetheweb) [inline sql/2024/fonts/design/fonts_metric.sql:125]: As discussed in chat would 100 be better than 99? It’s kinda meaningless at this extremes percentile anyway but highlighting the “worst use” can sometimes lead to interesting insights whereas I don’t think there’s anything of interest in p99 (since it’s highly volatile but also not that worse case).

File: sql/2024/fonts/development/fonts_color_color.sql [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+  return [];
+}
+''';
+
+WITH
+fonts AS (
+  SELECT
+    client,
+    url,
+    COLORS(JSON_EXTRACT(ANY_VALUE(payload), '$._font_details.color.palettes')) AS colors,
```
Comment (tunetheweb) [inline sql/2024/fonts/development/fonts_color_color.sql:34]: Why ANY_VALUE?

File: sql/2024/fonts/development/fonts_color_entry.sql [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+-- INCLUDE ../common.sql
+
+WITH
+fonts AS (
+  SELECT
+    client,
+    url,
+    SAFE_CAST(
+      JSON_EXTRACT_SCALAR(
+        ANY_VALUE(payload),
```
Comment (tunetheweb) [inline sql/2024/fonts/development/fonts_color_entry.sql:14]: Ditto

File: sql/2024/fonts/development/fonts_color_color.sql (hunk 18 above)
Comment (IvanUkhov) [reply sql/2024/fonts/development/fonts_color_color.sql:34]: So we group by the client and URL. For that combination, there are typically many entries corresponding to requests performed from different pages and at different times. It is then assumed the server's response does not depend on the location and time, and we just take any one here and in all other queries with the same grouping.

File: sql/2024/fonts/design/fonts_designer.sql (hunk 16 above)
Comment (IvanUkhov) [reply sql/2024/fonts/design/fonts_designer.sql:17]: We changed the crawler after the 2024-06-01 data were collected.

File: sql/2024/fonts/design/fonts_family_by_script.sql (hunk 15 above)
Comment (IvanUkhov) [reply sql/2024/fonts/design/fonts_family_by_script.sql:29]: BigQuery was running out of resources, and I was downsampling a little to make it go through. I have just rerun the query without the downsampling, and it succeeded. Perhaps it was something temporary. I have removed this code now.

File: sql/2024/fonts/common.sql (hunk 13 above)
Comment (IvanUkhov) [reply sql/2024/fonts/common.sql:63]: We have only one query dealing with JavaScript scripts (`performance/scripts_font_face.sql`). I think there will be little confusion what _script_ refers to when browsing the queries for the Fonts chapter. I am also assuming that in the article, one will be careful to explain.

File: sql/2024/fonts/design/fonts_designer.sql (hunk 16 above)
Comment (tunetheweb) [reply sql/2024/fonts/design/fonts_designer.sql:17]: @nrllh FYI, just making sure you're cool with this?

File: sql/2024/fonts/README.md (hunk 12 above)
Comment (IvanUkhov) [reply sql/2024/fonts/README.md:50]: Is it a hard rejection, or can this be discussed? We arguably want the queries to be easy to understand and easy to maintain. Duplicating the common functions at the top of each query will clutter the code and make it difficult to maintain, as one would have to keep track of the functions and keep them in sync.

If you want, we can replace

```sql
-- INCLUDE ../common.sql
```

with

```sql
-- INCLUDE https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2024/fonts/common.sql
```

Another alternative found be do create permanent SQL functions in a namespace in BigQuery dedicated to the chapter. Someone with the necessary permissions would have to do that.

File: sql/2024/fonts/design/fonts_metric.sql (hunk 17 above)
Comment (IvanUkhov) [reply sql/2024/fonts/design/fonts_metric.sql:125]: I think everything above the 90th will be ignored in the body of the chapter. I lowered it just to have something more sensible there; there will always be a handful of corky fonts with unrealistic parameters set.

File: sql/2024/fonts/development/fonts_color_color.sql (hunk 18 above)
Comment (tunetheweb) [reply sql/2024/fonts/development/fonts_color_color.sql:34]: Oh OK. I expected to look at fonts per site, but here we're looking at fonts overall (no matter which site requests them). OK.

File: sql/2024/fonts/README.md (hunk 12 above)
Comment (tunetheweb) [reply sql/2024/fonts/README.md:50]: I think that full path of the include would be a bit better.

**Review by tunetheweb — APPROVED** [review]

**Comment by tunetheweb** [comment]
Let me know when good to merge.

File: sql/2024/fonts/design/fonts_designer.sql (hunk 16 above)
Comment (nrllh) [reply sql/2024/fonts/design/fonts_designer.sql:17]: That's okay IMO

**Comment by IvanUkhov** [comment]
@tunetheweb, Bram and I concluded that it can be merged now. Thank you.

**Review by tunetheweb — APPROVED** [review]
